client/src/actions/sharedActions.js

```diff
@@ -1,29 +1,25 @@
 // @flow
-import produce from 'immer'
-import {get} from 'lodash'
+import {mappingFn as defaultMappingFn} from '../utils'
 
 import type {GenericAction, Path} from '../types/reduxTypes'
-import type {State} from '../state'
 
-// merges new data into destination Path
-// assumes a certain kind of data format - will need refinment if the API's different
+// merges new data into destination Path according to the mappingFn provided
 export const receiveData = (
   path: Path,
   data: Array<Object> | Object,
-  dataProviderRef: string
-): GenericAction<State, Array<Object> | Object> => ({
+  dataProviderRef: string,
+  mappingFn: (Array<Object> | Object, ...args?: Array<any>) => Object = defaultMappingFn,
+  ...mappingFnArgs: Array<any>
+): GenericAction<Object, Array<Object> | Object> => ({
   type: `Received data from ${dataProviderRef}`,
+  path,
   payload: data,
-  reducer: (state, data) =>
-    produce(state, (draft): void => {
-      const dataObject = Array.isArray(data)
-        ? data.reduce((obj, current) => {
-          obj[current.id] = current
-          return obj
-        }, {})
-        : {
-          [data.id]: data,
-        }
-      Object.assign(get(draft, path), dataObject)
-    }),
+  reducer: (state, data) => ({...state, ...mappingFn(data, ...mappingFnArgs)}),
+})
+
+export const updateValue = <T: *>(path: Path, data: T): GenericAction<T, T> => ({
+  type: 'Update data on path',
```

Comment:
I like this generic action, but I would like to have an option to set custom type... Also the default could be ``` `Update data on path ${path}` ```. But type would then need to be accept function...

Also, I tried using an invalid path, and unofortunately no error was raised :(. This is a a general problem of our actions though... I think it should check the existance of the state by default (although, you should be able to supress it as sometimes you want to put something on path which doesn't exist in state...). What do you think?

Reply:
The path is left out of message because it is logged just as well in the path part of object.

I'd leave this as is (apart from adding the optional custom type) - yes, getting something from state using a predefined path is not typed, but I'd just live with it, I don't think it's worth spending the extra effort to get the coverage.